### PR TITLE
Rajkumar42/myssh/osxlatest

### PR DESCRIPTION
**Pull request by rajkumar42**
Using lsof instead of readlink to get the full path of the executable.
Downloading GetClrDbg.sh on the MIEngine side and copying it over to the remote machine.
Cleaning up todos.

### CODE REVIEW

**Comment by rajkumar42** [comment]
@gregg-miskelly @caslan @chuckries @wesrupert

File: src/MICore/MICoreResources.resx [hunk 1]
```diff
@@ -243,4 +243,12 @@ Error: {1}</value>
   <data name="Info_InstallingDebuggerOnRemote" xml:space="preserve">
     <value>Installing debugger on the remote machine.</value>
   </data>
+  <data name="Info_StartingUnixCommand" xml:space="preserve">
+    <value>Starting unix command: '{0}'</value>
+    <comment>{0} command run on the remote unix machine.</comment>
+  </data>
+  <data name="Warning_DownloadingClrDbgToRemote" xml:space="preserve">
+    <value>Warning: Downloading Clrdbg to remote machine failed with error {0}.</value>
```
Comment (caslan) [inline src/MICore/MICoreResources.resx]: Maybe say .NET Core debugger instead of clrdbg?

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 2]
```diff
@@ -757,12 +771,10 @@ private void AddExecutablePathCommand(IList<LaunchCommand> commands)
...
         private void DetermineAndAddExecutablePathCommand(IList<LaunchCommand> commands)
         {
-            // TODO: rajkumar42, make it robust, either use the liblinux api or fallback shell commands available in other distros.
-            // TODO: rajkumar42, -target-attach can fail with gdb if ptrace>=1, elevate permission with liblinux and retry, display error on second failure.
-            // TODO: rajkumar42, connecting to OSX via SSH doesn't work yet.
+            // TODO: rajkumar42, connecting to OSX via SSH doesn't work yet. Show error after connection manager dialog gets dismissed.
 
             // Runs a shell command to get the full path of the exe.
-            string absoluteExePath = string.Format(CultureInfo.InvariantCulture, @"shell readlink -f /proc/{0}/exe", _launchOptions.ProcessId);
+            string absoluteExePath = System.FormattableString.Invariant($"shell lsof -p {_launchOptions.ProcessId} | awk '$4 == \"txt\" {{ print $9 }}'|awk 'NR==1 {{print $1}}'");
```
Comment (caslan) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: As discussed, add a comment about what the shell script does (e.g. that you are assuming the first open file will be the executable itself)

File: src/MICore/LaunchOptions.cs [hunk 3]
```diff
@@ -435,39 +435,26 @@ public sealed class UnixShellPortLaunchOptions : LaunchOptions
...
         /// </summary>
         /// /// <remarks>
-        /// {0} - DebuggerInstallationDirectory
-        /// {1} - GetClrDbg.sh url.
+        /// {0} - Base directory of debugger
+        /// {1} - clrdbg version.
         /// {2} - Subdirectory where clrdbg should be installed.
         /// </remarks>
-        private const string ClrdbgSubsequentLaunchCommand = "{0}/GetClrDbg.sh -v {1} -l {0}/{2} -d -s";
+        private const string ClrdbgSubsequentLaunchCommand = "pushd {0} && ./GetClrDbg.sh -v {1} -l {0}/{2} -d -s";
```
Comment (wesrupert) [inline src/MICore/LaunchOptions.cs]: Is there an associated `popd`, or is a `cd` sufficient?

File: src/MIDebugEngine/ResourceStrings.resx [hunk 4]
```diff
@@ -236,4 +236,12 @@
     <value>Failed to get executable path with error '{0}'.</value>
     <comment>0 = failure message</comment>
   </data>
+  <data name="Error_PTraceFailure" xml:space="preserve">
+    <value>Attaching to process {0} with {1} failed because of insufficient privilages with error message {2}.
```
Comment (caslan) [inline src/MIDebugEngine/ResourceStrings.resx]: do we want error message in quotes?

File: src/SSHDebugPS/Connection.cs [hunk 5]
```diff
@@ -47,5 +47,64 @@ internal void BeginExecuteAsyncCommand(string commandText, IDebugUnixShellComman
...
+            bool directoryExists = false;
+            liblinux.IO.IRemoteFileSystemInfo stat = null;
+            try
+            {
+                stat = _remoteSystem.FileSystem.Stat(path);
+                directoryExists = stat.IsDirectory();
+            }
+            catch
+            {
+                // Eat exceptions.
```
Comment (caslan) [inline src/SSHDebugPS/Connection.cs]: why?

File: src/MICore/LaunchOptions.cs [hunk 6]
```diff
@@ -513,13 +500,16 @@ public sealed class UnixShellPortLaunchOptions : LaunchOptions
                         startRemoteDebuggerCommand = "lldb-mi --interpreter=mi";
                         break;
                     case MIMode.Clrdbg:
+                        string userHomeDirectory = UnixPort.GetUserHomeDirectory();
+                        string debuggerHomeDirectory = System.FormattableString.Invariant($"{userHomeDirectory}{Path.AltDirectorySeparatorChar}{DebuggerInstallationDirectory}{Path.AltDirectorySeparatorChar}");
```
Comment (wesrupert) [inline src/MICore/LaunchOptions.cs]: Should probably use `String.Concat`

File: src/MICore/Transports/UnixShellPortTransport.cs [hunk 7]
```diff
@@ -21,7 +25,8 @@ public class UnixShellPortTransport : ITransport, IDebugUnixShellCommandCallback
         private bool _debuggerLaunched = false;
         private UnixShellPortLaunchOptions _launchOptions;
 
-        private static string s_errorPrefix = "Error:";
+        private const string s_errorPrefix = "Error:";
+        private const string ShellScriptName = "GetClrDbg.sh";
```
Comment (wesrupert) [inline src/MICore/Transports/UnixShellPortTransport.cs]: Inconsistent naming conventions

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 8]
```diff
@@ -643,8 +643,22 @@ private List<LaunchCommand> GetInitializeCommands()
...
+                        else
+                        {
+                            string message = string.Format(CultureInfo.CurrentUICulture, ResourceStrings.Error_ExePathInvalid, _launchOptions.ExePath, MICommandFactory.Name, miError);
+                            throw new LaunchErrorException(message);
+                        }
+                    };
+
                     int pid = _launchOptions.ProcessId;
-                    commands.Add(new LaunchCommand(String.Format(CultureInfo.CurrentUICulture, "-target-attach {0}", pid), ignoreFailures: false));
+                    commands.Add(new LaunchCommand(String.Format(CultureInfo.CurrentUICulture, "-target-attach {0}", pid), ignoreFailures: false, failureHandler: failureHandler));
```
Comment (wesrupert) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: InvariantCulture, or preferably just `"-target-attach " + pid` without all the formatting

Comment (wesrupert) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: Actually, just use `"-target-attach " + _launchOptions.ProcessId`, line 660 is extraneous.

**Comment by wesrupert** [comment]
As nice as they are, the formattable strings don't exist in the mono version that MI Engine builds against. Pretty sure this will break the "Desktop" build configurations.

**Comment by wesrupert** [comment]
LGTM! :shipit:

**Comment by chuckries** [comment]
:+1:

File: src/MICore/LaunchOptions.cs [hunk 9]
```diff
@@ -421,7 +421,7 @@ public sealed class UnixShellPortLaunchOptions : LaunchOptions
         /// <summary>
         /// Default location of the debugger on the remote machine.
         /// </summary>
-        public string DebuggerInstallationDirectory { get; private set; } = "~/.vs-debugger";
+        public string DebuggerInstallationDirectory { get; private set; } = ".vs-debugger";
```
Comment (gregg-miskelly) [inline src/MICore/LaunchOptions.cs]: >" [](start = 76, length = 1)

One thing we loose with this change is that it is no longer possible to install to a non-home-relative directory using the AD7 metric. I wonder if we should check if the first character of this is a '/' and if so assume that the path is rooted.

File: src/MICore/Transports/UnixShellPortTransport.cs [hunk 10]
```diff
@@ -34,12 +39,70 @@ public void Init(ITransportCallback transportCallback, LaunchOptions options, Lo
...
-            waitLoop?.SetText(MICoreResources.Info_InstallingDebuggerOnRemote);
+            if (_launchOptions.DebuggerMIMode == MIMode.Clrdbg)
+            {
+                if (!UnixShellPortLaunchOptions.HasSuccessfulPreviousLaunch(_launchOptions))
+                {
+                    waitLoop?.SetText(MICoreResources.Info_InstallingDebuggerOnRemote);
+                    string remoteDebuggerLocation = null;
+                    try
+                    {
+                        remoteDebuggerLocation = DownloadAndCopyFileToRemote(_launchOptions.DebuggerInstallationDirectory, _launchOptions.GetClrDbgUrl).Result;
```
Comment (gregg-miskelly) [inline src/MICore/Transports/UnixShellPortTransport.cs]: >remoteDebuggerLocation [](start = 24, length = 22)

This isn't used for anything. Do we want it?

File: src/MIDebugEngine/ResourceStrings.Designer.cs [hunk 11]
```diff
@@ -118,6 +118,17 @@ internal class ResourceStrings {
         }
         
         /// <summary>
+        ///   Looks up a localized string similar to Attaching to process {0} with {1} failed because of insufficient privilages with error message &apos;{2}&apos;.
+        ///
+        ///To attach to application on Linux, login as super user or disable ptrace..
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/ResourceStrings.Designer.cs]: >disable ptrace [](start = 69, length = 14)

We aren't actually disabling ptrace but rather allowing ptrace to be used across user accounts. 

I wonder if we want an aka.ms link for directions?

**Comment by gregg-miskelly** [comment]
>                 string message = string.Format(CultureInfo.CurrentUICulture, ResourceStrings.Error_FailedToGetExePath, miError);

Do you need a permission denied error handler here? I was seeing a failure with 'lsof -p 7' (which is my rcu_sched process)

---
Refers to: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs:781 in 31aac04. [](commit_id = 31aac04dfe0e1804f422e194814c3ebf8f06a3a7, deletion_comment = False)

File: src/SSHDebugPS/StringResources.resx [hunk 12]
```diff
@@ -123,9 +123,16 @@
...
   </data>
+  <data name="Error_InvalidDirectory" xml:space="preserve">
+    <value>The specified directory {0} could not be created or accessed.</value>
+    <comment>{0} directory name.</comment>
+  </data>
   <data name="Error_PSFailed" xml:space="preserve">
     <value>Unable to obtain process listing. 'ps' command failed.</value>
   </data>
+  <data name="Error_SourceFileNotFound" xml:space="preserve">
+    <value>Source file not found</value>
```
Comment (gregg-miskelly) [inline src/SSHDebugPS/StringResources.resx]: >not [](start = 23, length = 3)

Should we include the path to the file?

**Comment by gregg-miskelly** [comment]
Otherwise looks good to me

File: src/MICore/LaunchOptions.cs (hunk 9 above)
Comment (rajkumar42) [inline src/MICore/LaunchOptions.cs]: Agreed!

File: src/MICore/LaunchOptions.cs [hunk 13]
```diff
@@ -513,13 +500,24 @@ public sealed class UnixShellPortLaunchOptions : LaunchOptions
                         startRemoteDebuggerCommand = "lldb-mi --interpreter=mi";
                         break;
                     case MIMode.Clrdbg:
+                        string debuggerHomeDirectory;
+                        if (DebuggerInstallationDirectory.StartsWith("/", StringComparison.OrdinalIgnoreCase))
+                        {
+                            debuggerHomeDirectory = DebuggerInstallationDirectory;
```
Comment (gregg-miskelly) [inline src/MICore/LaunchOptions.cs]: >DebuggerInstallationDirectory [](start = 52, length = 29)

Should this add a trailing slash if it isn't already present?

File: src/MICore/Transports/UnixShellPortTransport.cs [hunk 14]
```diff
@@ -34,12 +39,69 @@ public void Init(ITransportCallback transportCallback, LaunchOptions options, Lo
...
 
-            waitLoop?.SetText(MICoreResources.Info_InstallingDebuggerOnRemote);
+            if (_launchOptions.DebuggerMIMode == MIMode.Clrdbg)
+            {
+                if (!UnixShellPortLaunchOptions.HasSuccessfulPreviousLaunch(_launchOptions))
+                {
+                    waitLoop?.SetText(MICoreResources.Info_InstallingDebuggerOnRemote);
+                    try
+                    {
+                        Task.Run(async () => await DownloadAndCopyFileToRemote(_launchOptions.DebuggerInstallationDirectory, _launchOptions.GetClrDbgUrl)).Wait();
```
Comment (gregg-miskelly) [inline src/MICore/Transports/UnixShellPortTransport.cs]: >async () => await [](start = 33, length = 17)

nit: you don't need the 'async' / 'await' part of this as DownloadAndCopyFileToRemote already returns a Task

File: src/MICore/LaunchOptions.cs (hunk 13 above)
Comment (rajkumar42) [inline src/MICore/LaunchOptions.cs]: That is not necessary, because it is gets added later on. Removed the redundant /.

**Comment by gregg-miskelly** [comment]
:shipit:

**Pull request merged**